Amazon S3 Media Storage support (#11738)
OrchardCMS/OrchardCore: src/docs/reference/modules/Media.AmazonS3/README.md

> # Amazon S3 Media Storage (`OrchardCore.Media.AmazonS3`)

Reviewer: Please document what kind of bucket configuration is necessary (like ACLs).

Media is still served by the Orchard Core website, and the Media Cache module takes responsibility for fetching media, on the fly, from the Amazon S3 Bucket.

This allows the Amazon Media Storage feature to support image resizing on the fly through integration with `ImageSharp.Web`.

The URL generated by the AssetUrl helpers points to the Orchard Core website.

## Configuration

The following configuration values are used by default and can be customized:

```json
{
   "OrchardCore": {
       "OrchardCore_Media_Amazon_S3": {
           // Your AWS S3 Bucket name
           "BucketName": "",
           // This section needed only if Orchard will be hosted not in the AWS Cloud
           // You can obtain all that information in the IAM Management Console
            "Credentials": {
              "SecretKey": "",
              "AccessKeyId": "",
              "RegionEndpoint": ""
           },
           // Optionally, set to a path to store media in a subdirectory inside your container.
           "BasePath": "/media",
           // If you have aws cli installed and configured you may just specify profile name
           "ProfileName": ""
       }
  }
}
```

> Refer also to the [Configuration Section](../../core/Configuration/README.md),
> and the [Media Section](../Media/README.md) for other Media related configuration settings.
> 
> There are two hosting options: inside and outside AWS Cloud.
> In case you are hosting Orchard Core inside AWS (EC2, EKS, etc.) you need to configure only `BucketName` and you may delete or comment out the other sections.
> 
> In case you are hosting Orchard Core outside of AWS, you should fill the `Credentials` section or if you have AWS CLI installed and configured on your server you may specify only configured profile name (`default` if a profile name was not chosen during AWS CLI configuration).
> 
> You can find region endpoints in the [Official AWS S3 Documentation](https://docs.aws.amazon.com/general/latest/gr/s3.html), see Region column. For example for Frankfurt region you should use `eu-central-1` 
> 
> ## AWS S3 Bucket Configuration
> 
> By default, AWS 3S Bucket has limitations for newly uploaded files. If you want media files to be available from the outside of AWS, you should set up your bucket permissions.
> 
> The simplest way of doing it is to add a policy:

Reviewer: ```suggestion
Refer also to the [Configuration Section](../../core/Configuration/README.md),
and the [Media Section](../Media/README.md) for other Media related configuration settings.

There are two hosting options: inside and outside AWS Cloud.
In case you are hosting Orchard Core inside AWS (EC2, EKS, etc.) you need to configure only `BucketName` and you may delete or comment out the other sections.

In case you are hosting Orchard Core outside of AWS, you should fill the `Credentials` section or if you have AWS CLI installed and configured on your server you may specify only configured profile name (`default` if a profile name was not chosen during AWS CLI configuration).

You can find region endpoints in the [Official AWS S3 Documentation](https://docs.aws.amazon.com/general/latest/gr/s3.html), see Region column. For example for the Frankfurt region you should use `eu-central-1` 

## AWS S3 Bucket Configuration

By default, AWS 3S Bucket has limitations for newly uploaded files. If you want media files to be available from the outside of AWS, you should set up your bucket permissions.

The simplest way of doing it is to add a policy:
```

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "AddPerm",
            "Effect": "Allow",
            "Principal": "*",
            "Action": "s3:GetObject",
            "Resource": "arn:aws:s3:::YOUR-BUCKET-NAME/*"
        }
    ]
}
```
After this policy will be added to your bucket permissions all newly added files will have Read permission and will be available from the outside of the Amazon Cloud.

## Templating Configuration

Optionally you may use liquid templating to further configure Amazon Media Storage, perhaps creating a bucket per tenant,
or a single bucket with a base path per tenant.

> The `ShellSettings` property is made available to the liquid template.

Reviewer: Capitalize Liquid everywhere.

> After this policy will be added to your bucket permissions all newly added files will have Read permission and will be available from the outside of the Amazon Cloud.
> 
> ## Templating Configuration
> 
> Optionally you may use liquid templating to further configure Amazon Media Storage, perhaps creating a bucket per tenant,
> or a single bucket with a base path per tenant.
> 
> The `ShellSettings` property is made available to the liquid template.
> The `BucketName` property and the `BasePath` property are the only templatable properties.
> 
> !!! note
> When templating the `BucketName`  using  `{{ ShellSettings.Name }}`, the tenant's name will be automatically lowercased, however, you must also make sure the `BucketName` conforms to other Amazon S3 naming conventions as set out in Amazon's documentation.

Reviewer: ```suggestion
After this policy will be added to your bucket permissions all newly added files will have Read permission and will be available from the outside of the Amazon Cloud.

## Templating Configuration

Optionally you may use liquid templating to further configure Amazon Media Storage, perhaps creating a bucket per tenant,
or a single bucket with a base path per tenant.

The `ShellSettings` property is made available to the liquid template.
The `BucketName` property and the `BasePath` property are the only templatable properties.

!!! note
    When templating the `BucketName` using `{{ ShellSettings.Name }}`, the tenant's name will be automatically lowercased, however, you must also make sure the `BucketName` conforms to other Amazon S3 naming conventions as set out in Amazon's documentation.
```

> ### Configuring a bucket per tenant.

Reviewer: ```suggestion
### Configuring a bucket per tenant
```

```json
{
    "OrchardCore": {
        "OrchardCore_Media_Amazon_S3": {
            "BucketName": "{{ ShellSettings.Name }}-media",
            "Credentials": {
                "SecretKey": "",
                "AccessKeyId": "",
                "RegionEndpoint": ""
            },
            "BasePath": "/media",
            "ProfileName": ""
        }
    }
}
```

> ### Configuring a single bucket, with a base folder per tenant.

Reviewer: ```suggestion
### Configuring a single bucket, with a base folder per tenant
```

## AWS S3 Bucket Configuration

> If `CreateBucket` was configured as `true` and `BucketName` is following official [Bucket naming rules](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html), than new bucket will be created.
> New bucket will be created without [Access Control Lists](https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html) due to security reasons, you may add needed ACLs manually:
> 1) Open you bucket
> 2) Go to the Permissions tab
> 3) Edit `Block public access` 

Reviewer: ```suggestion
If `CreateBucket` was configured as `true` and `BucketName` follows official [Bucket naming rules](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html), then a new bucket will be created.
The new bucket will be created without [Access Control Lists](https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html) due to security reasons. If you create the bucket manually then you need to do it with ACLs enabled. When using a previously created bucket, you may need to configure ACLs manually:

1. Open your bucket.
2. Go to the Permissions tab.
3. Edit "Block public access".
4. Tick "Block all public access".
```
